### PR TITLE
Out-of-memory exception handler

### LINKED ISSUE #9: Out-of-memory exception handler
Use exception handler for out of memory situations instead of returning NULL when memory allocation fails.

### PR DESCRIPTION
**Pull request by hanappe**
Close #9

### CODE REVIEW

File: src/membuf.c [hunk 1]
```diff
@@ -96,29 +86,26 @@ mutex_t *membuf_mutex(membuf_t *b)
...
+void membuf_append_zero(membuf_t *b)
 {
-        return membuf_put(b, 0);
+        membuf_put(b, 0);
 }
 
-int membuf_append_str(membuf_t *b, const char *s)
+void membuf_append_str(membuf_t *b, const char *s)
 {
         size_t lens = strlen(s);
```
Comment (tastymushroom) [inline src/membuf.c]: I think we were going to use strnlen here.

File: src/membuf.c [hunk 2]
```diff
@@ -197,7 +181,14 @@ int membuf_vprintf(membuf_t* b, const char* format, va_list ap)
         return -1;
 }
 
-int membuf_print_obj(membuf_t *b, json_object_t obj)
+static int32 membuf_json_writer(void* userdata, const char* s, int32 len)
+{
+        membuf_t *b = (membuf_t *) userdata;
```
Comment (tastymushroom) [inline src/membuf.c:178]: Should we check for NULL here before calling?

**Review by tastymushroom — COMMENTED** [review]
Hi Peter. 
Just two small comments for you to ponder.
Thanks a lot
Doug

File: src/membuf.c (hunk 1 above)
Comment (hanappe) [reply src/membuf.c]: Yes, I'll make the change.

File: src/membuf.c (hunk 2 above)
Comment (hanappe) [reply src/membuf.c:178]: json_serialise() passes the membuf 'b' to the callback membuf_json_writer() without changes. So would you reckon to add a single the test b!=NULL in membuf_print_obj(), or test for NULL in every callback to membuf_json_writer()?

Comment (tastymushroom) [reply src/membuf.c:178]: Hmm interesting.... It makes sense to have in in a single place so the membuf_print_obj() as, long as the writer isn't taken as a template in future and used elsewhere. Do you think we should harden membuf and check there in the append functions?

Comment (hanappe) [reply src/membuf.c:178]: OK. I will harden membuf and create a separate issue for that.
